cql3: Replace relation class with expression

cvybhu opened the pull request:
Before this change parser used to output instances of the `relation` class, which were later converted to `restrction`.
`relation` took care of initial processing such as preparing and some validation checks.

This PR aims to remove the `relation` class and perform it's functionality using only `expression`.
This is a step towards removing the legacy classes and converting all AST analysis to work on `expressions`.

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -20,15 +20,246 @@
+        case oper_t::GT:
+        case oper_t::GTE:
+            return true;
+        default:
```
It's better to list everything and avoid `default:` to avoid surprises when something new is added.

psarna commented:
> exceptions::syntax_exception (line 1:26 no viable alternative at input '>')

@cvybhu @havaker This one sounds like a problem with our "where clause". In materialized views (and hence, also secondary indexes) we keep the string representation of the where clause the view was created with. Ideally this where clause should not exist at all and we should just keep an expression in there, but that's a topic for a follow-up. Anyway, the where clause for secondary indexes is artificially created in the code, so perhaps you should investigate whether it always prepares valid CQL syntax:
https://github.com/scylladb/scylla/blob/d8c0680585b736a123698c46ccc8bafc44c58358/index/secondary_index_manager.cc#L146-L149

nyh commented:
> > exceptions::syntax_exception (line 1:26 no viable alternative at input '>')
> 
> @cvybhu @havaker This one sounds like a problem with our "where clause". In materialized views (and hence, also secondary indexes) we keep the string representation of the where clause the view was created with. Ideally this where clause should not exist at all and we should just keep an expression in there, but that's a topic for a follow-up. Anyway, the where clause for secondary indexes is artificially created in the code, so perhaps you should investigate whether it always prepares valid CQL syntax:

The specific test that @cvybhu mentioned (modify_table_with_view) has a view, not an index. Is this the only test failing with materialized views, or simply the first in the alphabetical order of source files? :-)  

Anyway, this test has a simple materialized view - "CREATE MATERIALIZED VIEW v AS SELECT * FROM t WHERE p > 0 PRIMARY KEY (p)". The where clause is "p > 0" but as @psarna hinted, the materialized view code does not actually keep this "p > 0" in string form - rather it parses it as usual, and later re-creates the string from the parsed form. I guess one of these two steps do not work as expected in the new expression code.

avikivity commented:
> > exceptions::syntax_exception (line 1:26 no viable alternative at input '>')
> 
> @cvybhu @havaker This one sounds like a problem with our "where clause". In materialized views (and hence, also secondary indexes) we keep the string representation of the where clause the view was created with. Ideally this where clause should not exist at all and we should just keep an expression in there, but that's a topic for a follow-up.

I've convinced myself the best serialization format for expressions is CQL.  Why spend more time devising a new serialization format, for mostly minor uses?

psarna commented:
> > > exceptions::syntax_exception (line 1:26 no viable alternative at input '>')
> > 
> > 
> > @cvybhu @havaker This one sounds like a problem with our "where clause". In materialized views (and hence, also secondary indexes) we keep the string representation of the where clause the view was created with. Ideally this where clause should not exist at all and we should just keep an expression in there, but that's a topic for a follow-up.
> 
> I've convinced myself the best serialization format for expressions is CQL. Why spend more time devising a new serialization format, for mostly minor uses?

Are we inventing a new serialization format? I only meant that the fact that we now store the `where` clause as a string, constructed artificially by first parsing CQL and then serializing it again manually, is neither optimal nor convenient. Instead, we should just store the `where` clause of a materialized view as an expression tree, because it is just that.

avikivity commented:
> 

It's stored in the schema, so it needs a serialization format. In memory it needs to be an expression for sure.

We also need to pass expressions for query pushdown.

psarna commented:
Ah, sure, storing it as CQL makes sense. If we ever need anything not expressable in CQL, we'll just add an internal extension

nyh commented on cql3/expr/expression.cc:
```diff
@@ -1087,7 +1088,7 @@ std::ostream& operator<<(std::ostream& os, const column_value& cv) {
     expression::printer pr {
         .expr_to_print = expr,
-        .mark_unresolved_identifiers = true
+        .debug_mode = true
```
please conflate this patch with previous patch, which added this flag with a different name

nyh commented on cql3/expr/expression.cc:
```diff
@@ -1109,7 +1109,7 @@ std::ostream& operator<<(std::ostream& os, const expression::printer& pr) {
             },
             [&] (const binary_operator& opr) {
-                if (pr.debug_mode) {
+                if (false) {
```
if (false)?

What's the purpose of this patch?

nyh commented on cql3/util.hh:
```diff
@@ -39,21 +39,45 @@ Result do_with_parser(const sstring_view& cql, Func&& f) {
+            .expr_to_print = e,
+            .debug_mode = false,
+        };
+        return fmt::format("{}", p);
```
When debug_mode=false, can't you just use format("{}", e) directly?

cvybhu replied:
No, the default is printing in debug mode, because that's how it was before.

nyh commented on cql3/util.hh:
```diff
@@ -39,21 +39,45 @@ Result do_with_parser(const sstring_view& cql, Func&& f) {
+        };
+        return fmt::format("{}", p);
+    };
+    auto expressions = relations | boost::adaptors::transformed(expr_to_pretty_string);
```
Is it possible that there are some operators with lower precedence than AND and you might need parentheses, or is this not a problem?

cvybhu replied:
AFAIK there aren't any - it's just AND and binary relations inbetween

nyh commented on cql3/expr/expression.hh:
```diff
@@ -610,7 +610,6 @@ extern expression search_and_replace(const expression& e,
         const noncopyable_function<std::optional<expression> (const expression& candidate)>& replace_candidate);
 
 extern expression prepare_expression(const expression& expr, data_dictionary::database db, const sstring& keyspace, lw_shared_ptr<column_specification> receiver);
-extern expression prepare_expression_multi_column(const expression& expr, data_dictionary::database db, const sstring& keyspace, const std::vector<lw_shared_ptr<column_specification>>& receivers);
```
The commit message is not descriptive enough. All I see here is code being deleted - but I don't know why (what it was used for before, and why it's no longer needed).

nyh commented on cql3/Cql.g:
```diff
@@ -153,30 +153,11 @@ struct uninitialized {
     bind_variable new_bind_variables(shared_ptr<cql3::column_identifier> name)
     {
-        auto marker = bind_variable{bind_variable::shape_type::scalar, _bind_variables.size()};
+        auto marker = bind_variable{_bind_variables.size()};
```
Again, for people less experienced in this code it would be good to expand a little bit on the commit message - what was this "shape_type" thing used for in the past, and why is it no longer needed?

nyh commented on test/cql-pytest/cassandra_tests/validation/entities/collections_test.py:
```diff
@@ -305,9 +305,9 @@ def testListWithUnsetValues(cql, test_keyspace):
         assert_rows(execute(cql, table, "SELECT l FROM %s WHERE k = 10"), [l])
 
         # select with in clause
-        assert_invalid_message(cql, table, "Invalid unset value for column k", "SELECT * FROM %s WHERE k IN ?", UNSET_VALUE)
```
Please don't remove tests - certainly not from cassandra_tests - without a thorough explanation of why. 
"WHERE k in UNSET_VALUE" *should* be an error. Why don't you want it to be an error?
They you moved it below the comment that says that the Python driver catches this case is surprising - did the Python driver change from the version I had to the one you have, and it's now caught the driver?

cvybhu replied:
The driver threw an error here on my branch, so I thought that maybe new driver version did some additional checks or something, but after closer investigation it turned out that it was because of my error It passes now without any modifications.

nyh commented on test/cql/query_bounds_test.result:
```diff
@@ -459,15 +459,21 @@ SELECT * FROM foo WHERE a=0 AND SCYLLA_CLUSTERING_BOUND(b) > SCYLLA_CLUSTERING_B
+			"b" : "0",
+			"c" : "0"
+		}
+	]
```
What's going on here? The test is called "too few values", and used to geenerate an error, but now suddenly doesn't generate an error. Why? Is it still "too few values"?

cvybhu replied:
Sending tuple with two values in place where tuple with three values is expected is sometimes correct.
For example if we send two `ints` to a `bind_variable` that has type `tuple<int, int, int>` Scylla will accept it and the third value will be `null`.

This is why validation code accepted this query - `tuple<int, int>` in place of `tuple<int, int, int>` is okay. I added checks for this specific case and now the test passes without any modifications.

cvybhu commented:
Alright the PR looks much better now and is almost ready to move out of draft.

All `dev-tests` pass and I didn't need to modify any tests in the end.

I will have another look at the code and maybe add some comments, but other than that everything is done.

The conditions for relation validation are taken from the old `relation.hh` code. I'm not sure how correct they are or whether they cover all cases, but I tried to keep the validation identical to how it was before.

cvybhu commented on cql3/expr/to_restriction.cc:
```diff
@@ -20,15 +20,246 @@
+
+bool is_legal_relation_for_non_frozen_collection(oper_t oper, bool is_lhs_col_indexed) {
+    return oper == oper_t::CONTAINS_KEY || oper == oper_t::CONTAINS || (oper == oper_t::EQ && is_lhs_col_indexed);
+}
```
I took this validation code from here:
https://github.com/scylladb/scylla/blob/063d26bc9edd7d0ad1b1a32c849cf46eae60b098/cql3/single_column_relation.cc#L143

I will take a closer look at it.

cvybhu replied:
> I think you should move this one line of code into the one place that uses this function.

This is gonna make the if look ugly :/

In case of CONTAINS KEY `prepare_binary_operator` will check that the type is a map and fail if it is not.

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -1100,49 +1101,90 @@ std::ostream& operator<<(std::ostream& os, const column_value& cv) {
+    };
+
     expr::visit(overloaded_functor{
             [&] (const constant& v) { os << v.view(); },
```
Argh the printer is printing `raw_value_view` here.
I think that's the hex representation of raw bytes :/
This might work for small ints that have the same hex version but other stuff probably breaks here.

psarna reviewed (commented):
Good job. I left a few minor comments

avikivity commented on cql3/expr/expression.hh:
```diff
@@ -218,10 +218,15 @@ const column_value& get_subscripted_column(const subscript&);
+/// Represents token(c1, c2) function on LHS of an operator relation.
+/// args contains arguments to the token function.
+struct token {
+    std::vector<expression> args;
```
Need updates to anything that treats nested expressions, like fill_prepare_context or search_and_replace.

cvybhu replied:
Damn I forgot about it, done.

psarna commented on cql3/expr/prepare_expr.cc:
```diff
@@ -415,7 +415,11 @@ list_prepare_expression(const collection_constructor& c, data_dictionary::databa
+            return evaluate_IN_list(value, query_options::DEFAULT);
+        } else {
+            return evaluate(value, query_options::DEFAULT);
+        }
```
hopefully https://github.com/scylladb/scylla/pull/10457 will drop the difference soon, and then we can purge evaluate_IN_list

cvybhu replied:
Sadly #10457 didn't remove the need for `evaluate_IN_list`.
It added support for bind markers inside collections, but there is still problem with `nulls`.
A list of `IN` values can contain nulls, but a normal collection can't.

avikivity replied:
`lists::setter::execute()` does not check for NULLs, so we likely have a bug in `UPDATE tab SET my_list = ?` if the bind variable contains a list with a NULL element.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -933,13 +916,7 @@ prepare_expression(const expression& expr, data_dictionary::database db, const s
-            case expr::bind_variable::shape_type::tuple_in: on_internal_error(expr_logger, "prepare_expression(bind_variable(tuple_in))");
-            }
-            on_internal_error(expr_logger, "unexpected shape in bind_variable");
+            return bind_variable_scalar_prepare_expression(bv, db, keyspace, receiver);
```
Very good change. It's like my previous comment regarding caller context in the callee: previously, bind_valiable_*_prepare_expression had to know about the caller context (communicated by calling different functions), and now the callee doesn't care and the caller configures it by setting the correct receiver type.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
 
+// Finds column_defintion for given column name in the schema.
+// Second argument is a relation to print in the error message in case of failure.
+static const column_definition* resolve_column(const unresolved_identifier& col_ident, const schema& schema, const binary_operator& printable_relation) {
```
Perhaps return a column_value here?

avikivity replied:
(so it's an expression, and compatible with the other stuff)

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+            return column_value(resolve_column(unin, schema, printable_relation));
+        },
+        [](const column_value& cv) -> expression { return cv; },
+        [&](const tuple_constructor& tc) -> expression {
```
Can this be part of the general prepare_expression()? Instead of just for binop lhs.

cvybhu replied:
It can't because we don't have a receiver for lhs

avikivity replied:
right, we need to make the receiver optional and a two-phase type inference pass.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+            const column_definition* sub_col = expr::visit(overloaded_functor {
+                [&](const unresolved_identifier& unin) -> const column_definition* {
+                    return resolve_column(unin, schema, printable_relation);
+                },
```
And here too, better to call prepare_expression recursively than to create new logic.

avikivity replied:
I think we need to add a `data_type` field to expressions. prepare_expression will fill this type.

avikivity replied:
Well, we have it in a weak form in get_lhs_receiver(). It can form the basis of more generalization later.

cvybhu replied:
`resolve_column` requires `schema`, which isn't passed as an argument to `prepare_expression`, but could probably be added.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+            return map_value_spec_of(*sub_col.col->column_specification);
+        },
+        [&](const tuple_constructor& tup) -> lw_shared_ptr<column_specification> {
+            sstring tuple_name = "(";
```
Please us std::ostringstream to avoid quadratic complexity.

cvybhu replied:
Done, I thought that strings work like vectors and allocate powers of 2, but they might not.

avikivity replied:
sstring does not, it always allocates the minimum amount. It did hit other people too.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+    lw_shared_ptr<column_specification> lhs_receiver = get_lhs_receiver(prepared_lhs, *schema);
+
+    lw_shared_ptr<column_specification> rhs_receiver = get_rhs_receiver(lhs_receiver, binop.op);
+    expression prepared_rhs = prepare_binop_rhs(binop.rhs, rhs_receiver, binop.op, db, schema->ks_name());
```
This can certainly be a recursive call to prepare_expression().

I guess prepare_binop_lhs() can't be, because we don't have a receiver. This is the core of type inference here: lhs has to be an expression around a column, so we can determine the type, then infer the rhs type by looking at the lhs type and the operator. 

Later, we can make prepare_expression behave in two ways:
1. with a receiver - validate types against receiver
2. without receiver - return a receiver, or error if cannot be determined (e.g. bind variable)

Then binary operators will first try preparing without a receiver, both sides, and use the returned receiver to prepare the other side.

cvybhu replied:
`prepare_binop_rhs` is basically a wrapper of `prepare_expression`, but it does one more check to detect and deal with lists of IN values which can contain nulls.

avikivity replied:
Let's see if we can move that check to runtime (and perhaps even: just ignore NULLs in IN lists, since no key can possibly match them).

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -20,15 +20,246 @@
                                                        prepare_context& ctx) {
+
+    if (is<constant>(e)) {
+        throw exceptions::invalid_request_exception("Constant value by itself is not supported as a restriction yet");
```
I definitely want to support it, not because it is useful, but because the grammar is a lot more flexible when a subexpression can be any expression, not just one of a few supported shapes.

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -20,14 +20,164 @@
+            binop_to_prepare.op = oper_t::EQ;
+            binop_to_prepare.rhs = elements[0];
+        }
+    }
```
Nice. We can move this to an optimizer, later.

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -20,14 +20,164 @@
+        }
+    }
+
+    binary_operator prepared_binop = prepare_binary_operator(binop_to_prepare, db, schema, ctx);
```
For later: we can call prepare_expression() here (and remove the restriction that it must be a binop). The receiver will have type boolean, since restrictions must evaluate to booleans. This will enable `WHERE bool_column` and the mighty `WHERE TRUE`.

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -20,14 +20,164 @@
+
+    if (prepared_binop.op == oper_t::NEQ) {
+        throw exceptions::invalid_request_exception(format("Unsupported \"!=\" relation: {}", prepared_binop));
+    }
```
Wait, we don't support this?

cvybhu replied:
We don't
```cql
cqlsh> CREATE KEYSPACE IF NOT EXISTS ks WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor' : 1};
cqlsh> USE ks;
cqlsh:ks> CREATE TABLE tab (p int primary key);
cqlsh:ks> SELECT * FROM tab WHERE p != 0 ALLOW FILTERING;
InvalidRequest: Error from server: code=2200 [Invalid query] message="Unsupported "!=" relation: p != 0"
```
Neither does cassandra

Here's the code responsible for it:
https://github.com/scylladb/scylla/blob/528ab5a5026e83c35a6a3b6d476230b692099a2a/cql3/relation.hh#L134

avikivity replied:
So it's impossible to use it anywhere, and it's just in the grammar?

I don't mind, but it's odd.

cvybhu replied:
> So it's impossible to use it anywhere, and it's just in the grammar?

It looks like there is one place in `column_condtion` that has an if for something like `a != NULL`.
https://github.com/scylladb/scylla/blob/528ab5a5026e83c35a6a3b6d476230b692099a2a/cql3/column_condition.cc#L190
But that's the only actually possible use of this operator that I found.
Btw shouldn't this be `a IS NOT NULL`?

`is_satisfied_by` already has implementation for `!=`, so it would be probably easy to add support for it, at least with allow filtering But maybe it's better to wait until the end of expression rewrite and then something more efficient could be implemented.

avikivity replied:
> > So it's impossible to use it anywhere, and it's just in the grammar?
> 
> It looks like there is one place in `column_condtion` that has an if for something like `a != NULL`.
> 
> https://github.com/scylladb/scylla/blob/528ab5a5026e83c35a6a3b6d476230b692099a2a/cql3/column_condition.cc#L190
> 
> 
> But that's the only actually possible use of this operator that I found.
> Btw shouldn't this be `a IS NOT NULL`?

Probably. Please add an issue.

> `is_satisfied_by` already has implementation for `!=`, so it would be probably easy to add support for it, at least with allow filtering But maybe it's better to wait until the end of expression rewrite and then something more efficient could be implemented.

Sure, it's low priority in any case.

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -20,14 +20,164 @@
+
+    if (prepared_binop.op == oper_t::IS_NOT && !is<null>(prepared_binop.rhs)) {
+        throw exceptions::invalid_request_exception(format("Unsupported \"IS NOT\" relation: {}", prepared_binop));
+    }
```
I think we should parse IS NULL and IS NOT NULL into unary operators, not binary operators with a NULL operand. But that can wait.

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -123,6 +125,156 @@ ::shared_ptr<restrictions::restriction> new_subscripted_column_restriction(subsc
+            return ::make_shared<restrictions::multi_column_restriction::IN_with_values>(schema,
+                                                                                         std::move(lhs_cols),
+                                                                                         std::move(rhs_list->elements));
+        }
```
These look very similar, I hope they are merged later.

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -123,6 +125,156 @@ ::shared_ptr<restrictions::restriction> new_subscripted_column_restriction(subsc
+            throw exceptions::invalid_request_exception("LIKE cannot be used for Multi-column relations");
+        }
+
+        if(auto rhs_tup = as_if<tuple_constructor>(&binop.rhs)) {
```
`if` is not a function.

cvybhu replied:
I don't understand what's wrong here

avikivity replied:
function: `f(...)`
not-function: `if (...)`

cvybhu replied:
You mean that there should be a space after if?
I added a space

avikivity replied:
yes

avikivity commented:
Very good, nice progress.

psarna commented:
@cvybhu since https://github.com/scylladb/scylla/pull/10457 made it upstream, I guess this series can also take advantage of it and simplify all the special cases for IN_list?

cvybhu commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1057,6 +1058,182 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+                tuple_types.push_back(elem_receiver->type->underlying_type());
+            }
+
+            tuple_name += ")";
```
A join here would be awkward because I do some other processing in the logic, `ostringstream` seems elegant enough for me.

avikivity replied:
> > What suggestions?
> 
> All of them apart from `evaluate_IN_list`. 

In the future, please keep a list somewhere while working and then add it to the update. Now I have to go look at the suggestions.

> I have a feeling that eliminating it will make the code uglier, but I will look and see what can be done.

I feel it will be prettier, but better to do it post-merge (if we decide to do it).

cvybhu commented:
v1:
* Fixed `expr::printer` and added some tests for it
* Applied suggestions from comments, still thinking whether to move logic of `evaluate_IN_list` somewhere higher in the call stack

I don't see any major issues with this PR anymore and it's ready to move out of draft.

avikivity commented:
> v1:
> * Applied suggestions from comments, still thinking whether to move logic of `evaluate_IN_list` somewhere higher in the call stack

What suggestions?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/365/

cvybhu commented:
> What suggestions?

All of them apart from `evaluate_IN_list`. I have a feeling that eliminating it will make the code uglier, but I will look at see what can be done.

cvybhu commented:
Damn it there is another issue with the printer

There is something funny going on with local/utc time when inserting timestamps and dates.
I have a test like this and it fails if time zone information is removed (the `+0000`).
```c++
BOOST_AUTO_TEST_CASE(expr_printer_timestamp_test) {
    constant timestamp_const (
        raw_value::make_value(timestamp_type->from_string("2011-03-02T03:05:00+0000")),
        timestamp_type
    );
    BOOST_REQUIRE_EQUAL(expr_print(timestamp_const), "'2011-03-02T03:05:00'");
}
```
Without timezone information the timestamp is interpreted as local time I think (?) because the printed value is a few hours off.

This is a problem because timestamp printer doesn't output timezone information, so parsing the output back will generate a different timestamp depending on the timezone. I could modify it to always append `+0000` to the output, but I want to investigate a bit first.
I find it really weird that timestamps use local time by default, it seems intuitive to do everything as UTC unless specified otherwise. Using local time of a database node somewhere random in the world could be really confusing 0_o

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/366/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/367/

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -0,0 +1,34 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
We changed the license description to SPDX, see for example https://github.com/scylladb/scylla/blob/master/main.cc#L6

cvybhu replied:
Changed to SPDX

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -0,0 +1,34 @@
+                                                       schema_ptr schema,
+                                                       prepare_context& ctx) {
+    throw std::runtime_error(fmt::format("Unimplemented! {}:{}", __FILE__, __LINE__));
+}
```
Please fold the implementing patch and this one into each other.

cvybhu replied:
done

avikivity commented on cql3/expr/expression.hh:
```diff
@@ -218,10 +217,15 @@ const column_value& get_subscripted_column(const subscript&);
-struct token {};
+    explicit token(std::vector<expression>);
+    explicit token(const std::vector<const column_definition*>&);
+    explicit token(const std::vector<::shared_ptr<column_identifier_raw>>&);
```
Do we need the last two constructors? The callers can convert their inputs to expressions.

If we remove them, we can remove the first constructor too.

avikivity replied:
(there's no reason for token to presume anything about its parameters)

avikivity replied:
we can have helpers make_token_expr_from(...) to ease the transition. Eventually the grammar will pass expressions there anyway.

cvybhu replied:
I feel like constructors are nicer than forcing the caller to create a vector of expressions.
With constructors the caller doesn't care what is the exact representation that token requires, they can just create it and move on. Converting is then neatly compartmentalized inside the constructor and caller can focus on their logic.
Additionally this eliminates repetition when other code will want to do the same thing.

I can add `make_token_expr_from(...)`, but is there any reason to have it instead of constructors? With constructors it's immediately obvious what they do and the function is easy to find for someone exploring the code. I feel like a free standing function would make it more complicated.

avikivity replied:
My reasoning is to have a struct-like approach where it's plain data. But it's a minor point, we can keep it as is.

avikivity replied:
Also I'll secretly hope that they become outdated when the grammar emits expressions for the arguments.

avikivity commented on cql3/single_column_relation.cc:
```diff
@@ -58,6 +58,26 @@ single_column_relation::new_EQ_restriction(data_dictionary::database db, schema_
+                              data_dictionary::database db,
+                              const sstring& keyspace,
+                              prepare_context& ctx) {
+    assert(receivers.size() == 1);
```
on_internal_error() is safer, it does not crash the server on a bug.

cvybhu replied:
Changed

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1046,6 +1047,183 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+        return list_prepare_expression(rhs_in_list, db, keyspace, rhs_receiver, true);
+    }
+
+    return prepare_expression(rhs, db, keyspace, rhs_receiver);
```
I think this highlights the bug in lists::operations::setter. In general prepare_expression checks for NULLs in collection_constructor, but there is no such check in collections received via bind variables. Or maybe there is and it is well hidden.

cvybhu replied:
You're right, there doesn't seem to be a proper check for null in a collection received from bind marker.
`evaluate(const bind_variable&)` Calls `validate` on the type, but `validate` doesn't check for nulls.
A check could be added in `validate()`.
In case of lists there could theoretically be nulls in case of lists of IN values, but this doesn't work at the moment so maybe we don't need to support it at all.

Actually now that I checked it cassandra doesn't allow for NULLs in an IN list anymore :O
We could forbid nulls there and we are still gonna be compatible with cassandra.
This will solve the `evaluate_IN_list` problem altogether :)
I will open an issue for it.

Trying to send a collection with null values inside fails but in an ugly way.
`validate()` does `read_collection_value` for all elements, which reads element size, interprets it as unsigned int and reads that much. In case of `null` value this size is `-1` so it tries to read some random huge number of bytes and then fails.

cvybhu replied:
Opened #10579 and #10580

avikivity commented on cql3/expr/to_restriction.cc:
```diff
@@ -20,15 +20,246 @@
+
+        check_false(lhs_col_type.is_collection(), "Slice restrictions are not supported on collections containing durations");
+        check_false(lhs_col_type.is_tuple(), "Slice restrictions are not supported on tuples containing durations");
+        check_false(lhs_col_type.is_user_type(), "Slice restrictions are not supported on UDTs containing durations");
```
What was the resolution here?

avikivity commented on cql3/Cql.g:
```diff
@@ -71,9 +71,6 @@ options {
-#include "cql3/token_relation.hh"
 #include "cql3/index_name.hh"
 #include "cql3/cql3_type.hh"
 #include "cql3/cf_name.hh"
```
For the patch subject line: add 'for relations', since it already outputs expressions elsewhere.

cvybhu replied:
done

avikivity commented:
Very good, I had a few minor requests.

cvybhu commented:
v2:
* Fixed printer for timestamp and date
* Folded to_restriction declaration and implementation into a single commit
* Changed assert to on_internal_error
* Changed commit name to `cql3: Make parser output expression for relations`
* Changed license notice to SPDX

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/378/

cvybhu commented on cql3/expr/prepare_expr.cc:
```diff
@@ -383,7 +383,7 @@ list_test_assignment(const collection_constructor& c, data_dictionary::database
 static
 expression
-list_prepare_expression(const collection_constructor& c, data_dictionary::database db, const sstring& keyspace, lw_shared_ptr<column_specification> receiver) {
+list_prepare_expression(const collection_constructor& c, data_dictionary::database db, const sstring& keyspace, lw_shared_ptr<column_specification> receiver, bool is_in_list) {
```
Yes, once they go in there will be no case in which nulls are allowed inside collections and this parameter will disappear
